src/dcap_provider.cpp

```diff
@@ -518,35 +578,170 @@ static sgx_plat_error_t build_pck_crl_url(
 static std::string build_tcb_info_url(
     const sgx_ql_get_revocation_info_params_t& params)
 {
-    std::string tcb_info_url =
-        get_base_url() + "/tcb/" +
-        format_as_hex_string(params.fmspc, params.fmspc_size) + '?';
+    std::string fmspc((char *)params.fmspc, params.fmspc_size);
+    return build_tcb_info_url(fmspc);
+}
 
+//
+// The expected URL for QeID or QveID
+//
+static std::string build_enclave_id_url(bool qve, std::string &expected_issuer_chain_header)
+{
+    std::string version = get_collateral_version();
     std::string client_id = get_client_id();
+    std::stringstream qe_id_url;
+    expected_issuer_chain_header = headers::QE_ISSUER_CHAIN;
 
+    qe_id_url << get_base_url();
+    if (!version.empty())
+    {
+        qe_id_url << "/" << version;
+        if (version != "v1")
+        {
+            expected_issuer_chain_header = headers::ENCLAVE_ID_ISSUER_CHAIN;
+        }
+    }
+     qe_id_url << "/" << (qve ? "qveid" : "qeid") << "?";
+    
     if (!client_id.empty())
     {
-        tcb_info_url += "clientid=" + client_id + '&';
+        qe_id_url << "clientid=" << client_id << '&';
     }
+    qe_id_url << API_VERSION;
+    return qe_id_url.str();
+}
 
-    return tcb_info_url + API_VERSION;
+static quote3_error_t get_collateral(std::string url, 
+                                            const char header_name[], 
+                                            std::vector<uint8_t> &body, 
+                                            std::string &issuer_chain)
+{
+    try 
+    {
+        const auto crl_operation = curl_easy::create(url);
+        crl_operation->perform();
+        body = crl_operation->get_body();
+        auto get_header_operation = get_unescape_header(*crl_operation,header_name, &issuer_chain);
+        // Insert null characters
+        body.push_back(0);
+        char nullcharacter = '\0';
+        issuer_chain.append(1, nullcharacter);

```

Comment:
To preserve whatever old behavior exists, I need to leaving get_unescape_header unchanged. C++ strings do not use the null character for string termination so IMO the cleanest way to get it into the C-style buffers is to add it here and return the std::string .data(). Calling .c_str() will add the null character but .size() would be off by one

Reply:
That's not technically true. It's certainly true that C++ strings are counted length strings and .size() does not include the terminating null, but the null *is* included in the internal representation of the string (this is required because string.data() returns a pointer to the underlying representation of the string including the null terminator.

Reply:
Thanks good to know! I still think I like adding the extra null characters here

Reply:
By adding the trailing nul, you're generating a string which is likely to be considered corrupt by most clients. Callers see a std::string and they assume that the string has only a single trailing nul, but your string will have two trailing nul characters and that may seriously confuse clients.

For body, since it's a std::vector<uint8_t>, adding the trailing nul is also inappropriate, because the body isn't always a string - the only scenario in which it would be appropriate to add the trailing nul is if the content-type: header indicates that it is text (Content-Type: text/<something>.

Otherwise you're corrupting the body.

Reply:
Ah understood I'll fix it

Reply:
This has been fixed - see fill_qpl_string_buffer in latest